wasm/derive/src/lib.rs

```diff
@@ -81,29 +82,229 @@ pub fn entrypoint(attr: TokenStream, item: TokenStream) -> TokenStream {
 
 /// Annotate the user-defined function that starts the execution of a validator.
 ///
-/// Annotated function should have one parameter of the type which implements
-/// `TryFrom<NeedsPermissionBox>`.
-///
-/// Validators are only checking if an operation is **invalid**, not its validness.
+/// Validators are only checking if an operation is **invalid**, not if it is valid.
 /// A validator can either deny the operation or pass it to the next validator if there is one.
 ///
+/// # Attributes
+///
+/// This macro must have an attribute describing entrypoint parameters.
+///
+/// The syntax is:
+/// `#[iroha_wasm::validator_entrypoint(params = "[<type>,*]")]`, where `<type>` is one of:
+/// - `authority` is a signer account id who submits an operation
+/// - `transaction` is a transaction that is being validated
+/// - `instruction` is an instruction that is being validated
+/// - `query` is a query that is being validated
+/// - `expression` is an expression that is being validated
+///
+/// Exactly one parameter of *operation to validate* kind must be specified.
+/// `authority` is optional.
```

Comment:
Hm... So here are my thougths on this:

1. We can't do inference on macro level, only on generated code level. This is because we can't rely on type name, but can rely on generated `try_from()` usage. This way user won't be able co compile invalid entrypoint, but we will not be able to provide a good error message and suggestion to use attributes.

2. If we will do this then we also need this for *triggers*, because they work on the same way right now.

3. In most cases you still need `authority`, so in most cases you will use this attributes, because we can't infere this as we don't know the exact order of parameters and still can't rely on type name

4. Will be tricky to implement

Reply:
I'm with @mversic on this one. Re: comments
1. Not an issue. Name based inference is called nominative typing. You will need provisions to disambiguate, but if you have a name collision like this one in Rust you're just as much in trouble. If you want to be extra careful, you could require this macro to accept an entire `mod` so that renaming declarations are stopped at compile time. 
2. Not a bad idea. Triggers could use some syntactic sugar. 
3. We can rely on the mapping `name: Type` to give us the full info. The `_authority` tells you volumes. 
4. Sure. So we might want to postpone it. But from an ergonomic point of view, it makes sense to invest some time before releasing `2.0.0`

Reply:
Okay, created #3260